### ISSUE TITLE
Implement function to execute a custom action

### ISSUE THREAD

**Issue by brentd313** [uchicago-cs/chiventure#705]
We will be working on implementing a custom action execution function. We will begin by reducing the overall function into more manageable component helpers. Ideally, the final function should be capable of accepting an arbitrary custom action structure as input and executing it appropriately; however, if that proves infeasible, achieving an implementation capable of executing a limited range of custom actions would provide a good foundation for future enhancement. We will keep this issue updated with the breakthroughs and obstacles we face.

**Comment by sonkinss** [comment]
Function run_ast_block(AST_block_t *block) created and implemented with a switch statement to execute the appropriate block type. Added do_branch_block IFELSE functionality, but need to implement do_conditional_block before we can use branch blocks.

Current issue: need to reevaluate how condition_block attributes will be stored. Doesn't appear that "player has item X" can currently be processed with our conditional block implementation, which is one of the most important ones

Additionally, will need to add function to execute a custom action starting from our wrapper struct of custom_action_t.

**Comment by brentd313** [comment]
I opened up a new branch titled custom-actions/exec-action-block to implement the function for executing an action block within a custom action. Similar to Sergey's comment on conditional blocks, the effect-applying functionalities currently supported by our code limit the range of actions that can be executed. However, the implementation should be flexible as support for new types of actions (aka effects) is added.

**Comment by brentd313** [comment]
The action block execution function is complete with unit tests. The code is compiling, and the tests are passing. You can find the tests in the test_action_block.c file in the tests/custom-actions directory.

**Comment by brentd313** [comment]
To provide some clarification on the approach we are taking to write the overall custom actions execution function:

A custom action structure is a wrapper struct that contains a linked list of "blocks" (component data structures). These blocks can be of type conditional, action, branch, or control, which are each placed in a wrapper "ast" block structure, which contains a union of the blocks, a tag and a pointer to the next block. We are creating helper functions that execute each type of block, and the final custom action execution function will be a loop that iterates through the linked list of blocks and uses a switch statement to execute the function appropriate for each block.

**Comment by brentd313** [comment]
The pull request with the action block execution code has been linked to this issue.

**Comment by brentd313** [comment]
The pull request for the action block execution has been merged into dev. If you're interested, the code related to this pull request can be found in action_block.c/.h in the src/custom-actions directory, and the tests can be found in the test_action_block.c file in the tests/custom-actions directory.

**Comment by LapsededPacifist** [comment]
Moved the function for executing an AST_block into ast_block.c and for evaluating a conditional into conditional_block.c to be consistent with action blocks. The tests for these functions will be found in their respective folders. 
Also changed the conditional enum away from LTGT and LTEGTE since that did not allow for differentiation when evaluating a conditional. This was instead changed to, LTB, GTB, LTEB, and GTEB to avoid conflict with the already existing LT, GT, LTE, and GTE enum.

**Comment by LapsededPacifist** [comment]
Final additions to custom actions:

-`run_ast_block` was added to ast_block.c/.h
-  - Executes an ast block. Can only run action blocks currently, but branch blocks will be added in [Issue 1068](https://github.com/uchicago-cs/chiventure/issues/1068). This is designed specifically for running an action, so it should not contain any conditional or control blocks. Runs the exec_action_block function on found action blocks.
- `Conditional_type_t` was modified to have LTB, GTB, LTEB, and GTEB instead of LTGT and LTEGTE . 
-  - Allows for eval_conditional_block to differentiate between less than and greater than.
- `Eval_conditional_block` was added to conditional_block.c/.h, but is waiting on do_branch_block to be implemented. Simply compares the two attributes in a conditional_block and returns true or false.
- `do_custom_action` was added to interface.c/.h. 
- - Just a wrapper for running `run_ast_block` on the `AST_block_t` list of a custom action
Note:
Also discovered errors in attribute new that are described in [issue 1072](https://github.com/uchicago-cs/chiventure/issues/1072). The function only seems to be running currently because the last function called is malloc, which means the pointer they want to return is likely still going to be in the return register.

**Comment by MaxineK36** [comment]
Issue Score: ✔️++

Comments:
Great job taking over this existing issue! Your updates are particularly detailed and useful.
